ibis/backends/pandas/executor/analytic.py

```diff
@@ -0,0 +1,106 @@
+from __future__ import annotations
+
+import numpy as np
+import pandas as pd
+
+import ibis.expr.operations as ops
+from ibis.backends.pandas.executor.core import execute
+
+
+@execute.register(ops.Lag)
+@execute.register(ops.Lead)
+def execute_lead_lag(op, arg, offset, default):
+    if isinstance(op, ops.Lag):
+        sign = lambda x: x
+    else:
+        sign = lambda x: -x
+
+    if op.offset is not None and op.offset.dtype.is_interval():
+
+        def agg(df, order_keys):
+            df = df.set_index(order_keys)
+            col = df[arg.name].shift(freq=sign(offset))
+            return col.reindex(df.index, fill_value=default)
+    else:
+        offset = 1 if offset is None else offset
+
+        def agg(df, order_keys):
+            return df[arg.name].shift(sign(offset), fill_value=default)
+
+    return agg
+
+
+@execute.register(ops.MinRank)
+@execute.register(ops.DenseRank)
+def execute_min_rank(op):
+    method = "dense" if isinstance(op, ops.DenseRank) else "min"
+
+    def agg(df, order_keys):
+        if len(order_keys) == 0:
+            raise ValueError("order_by argument is required for rank functions")
+        elif len(order_keys) == 1:
+            s = df[order_keys[0]]
+        else:
+            s = df[order_keys].apply(tuple, axis=1)
+
+        return s.rank(method=method).astype("int64") - 1
+
+    return agg
+
+
+@execute.register(ops.PercentRank)
+def execute_percent_rank(op):
+    def agg(df, order_keys):
+        if len(order_keys) == 0:
+            raise ValueError("order_by argument is required for rank functions")
+        elif len(order_keys) == 1:
+            s = df[order_keys[0]]
+        else:
+            s = df[order_keys].apply(tuple, axis=1)
+
+        return s.rank(method="min").sub(1).div(len(df) - 1)
+
+    return agg
+
+
+@execute.register(ops.CumeDist)
+def execute_cume_dist(op):
+    def agg(df, order_keys):
+        if len(order_keys) == 0:
+            raise ValueError("order_by argument is required for rank functions")
+        elif len(order_keys) == 1:
+            s = df[order_keys[0]]
+        else:
+            s = df[order_keys].apply(tuple, axis=1)
+
+        return s.rank(method="average", pct=True)
+
+    return agg
+
+
+@execute.register(ops.RowNumber)
+def execute_row_number(op):
+    def agg(df, order_keys):
+        return pd.Series(np.arange(len(df)), index=df.index)
+
+    return agg
+
+
+@execute.register(ops.FirstValue)
+@execute.register(ops.LastValue)
+def execute_first_value(op, arg):
+    i = 0 if isinstance(op, ops.FirstValue) else -1
+
+    def agg(df, order_keys):
+        return df[arg.name].iloc[i]
```

Comment:
```suggestion
        return df[arg.name].iat[i]
```

Supposedly `iat` is faster than `iloc` due to only handling scalar integer inputs (no arrays/sequences are allow)

Reply:
Updated.